### ISSUE TITLE
Expose Chromium content switches

### ISSUE THREAD

**Issue by The-Compiler** [qutebrowser/qutebrowser#2377]
Chromium's [content_switches.cc](https://cs.chromium.org/chromium/src/content/public/common/content_switches.cc?q=content_switch+package:%5Echromium$&l=1) has some interesting options we could expose as options which require a restart.

**Comment by throwaway217** [comment]
So I can set stuff from content_switches.cc with qt-flags like:
--qt-flags='aec-refined-adaptive-filter, disable-accelerated-video-decode' ?
Some of these look like they're already available due to webengine. What does expose actually mean ?

**Comment by The-Compiler** [comment]
Not like that, but `--qt-flag aec-refined-adaptive-filter --qt-flag disable-accelerated-video-decode` should work. Note not all of them will work.

With expose I mean as qutebrowser settings. Many of them are probably too advanced, but things like `disable-reading-from-canvas` (#2235) or maybe `--disable-gpu` (#2368) would be nice to have as a setting.

The first step would be finding out which ones actually work, and then from that list pick the ones which an average qutebrowser user might like to customize.

**Comment by throwaway217** [comment]
>Note not all of them will work.
Am I getting the right image ? Do these have to be tested manually every webengine release or not all of them work because they weren't implemented yet ?

**Comment by The-Compiler** [comment]
Ideally we'd have automated tests for the ones we want to expose. Not sure how much will change with new QtWebEngine releases.

**Comment by The-Compiler** [comment]
https://codereview.qt-project.org/#/c/200248/1//ALL,unified (probably part of Qt 5.9.2) adds support for `--force-webrtc-ip-handling-policy` which can be used to prevent WebRTC IP leaks.

**Comment by The-Compiler** [comment]
There's a `qt_args` setting in `master` - I don't think the other settings need to be in v1.0.0.

**Comment by The-Compiler** [comment]
I added some more switches, and don't think anything else from https://peter.sh/experiments/chromium-command-line-switches/ should be exposed.

However, it's possible that I missed something, and looking through those is a nice way to help qutebrowser without needing to be able to code.

chrome://flags/ inside Chrome/Chromium is also a nice place to look, many of those things could be exposed to qutebrowser as well.

**Comment by rumpelsepp** [comment]
is the quic protocol included/available in QT? Would be a notable improvement, especially for roaming users.

**Comment by rumpelsepp** [comment]
> Hard to say without any demo server (or server which already speaks QUIC?)

All google servers already speak quic for quite some time; including youtube GMail, etc.

> That seems unrelated to commandline switches. 

Oops. Sorry, the title just speaks about "expose more switches".

**Comment by The-Compiler** [comment]
> All google servers already speak quic for quite some time; including youtube GMail, etc.

I don't see any QUIC/UDP in Wireshark when visiting Google servers in Chromium.

**Comment by rumpelsepp** [comment]
On linux you have to enable it in `chrome:settings`

**Comment by The-Compiler** [comment]
Ah, right! Looks like there's indeed an `--enable-quic` flag (even a commandline flag for Chromium), but passing `--qt-flag enable-quic` to qutebrowser (Qt 5.12, Chromium 69.0.3497.128) doesn't seem to do anything.

**Comment by rumpelsepp** [comment]
Nice! Thanks.

**Comment by livnev** [comment]
@The-Compiler Could you clarify if it is now possible to invoke some equivalent of `--disable-gpu`? On my setup this is the only way I can run chromium, and qutebrowser is having the same issue. So far, I've been able to get it working with `LIBGL_ALWAYS_SOFTWARE=1 qutebrowser`.

**Comment by The-Compiler** [comment]
@livnev Sounds like you want the `qt.force_software_rendering` setting.

**Comment by livnev** [comment]
@The-Compiler thanks, `c.qt.force_software_rendering = "software-opengl"` is what I was looking for.

**Comment by The-Compiler** [comment]
I just looked through some of the [Blink settings](https://cs.chromium.org/chromium/src/third_party/blink/renderer/core/frame/settings.json5) - most of them don't work, but there's a group of "force dark mode" settings there which work (since Qt 5.14) and which I'd like to expose.

Relevant links:

- [Settings definitions](https://cs.chromium.org/chromium/src/third_party/blink/renderer/core/frame/settings.json5?q=darkModeImagePolicy&sq=package:chromium&dr=C&l=876)
- [Related issue](https://bugs.chromium.org/p/chromium/issues/detail?id=1002664)
- [Issue with some example configs](https://bugs.chromium.org/p/chromium/issues/detail?id=985961)

## darkMode

Qt 5.14 way of setting the dark mode algorithm. In the current Chromium, this is renamed to `darkModeInversionAlgorithm`.

- kOff = 0 (deprecated)
- kSimpleInvertForTesting = 1 (For testing only, does a simple 8-bit invert of every RGB pixel component)
- kInvertBrightness = 2
- kInvertLightness = 3
- kInvertLightnessLAB = 4 (default)

## darkModeEnabled

bool, instead of the above in the current Chromium version, so probably in Qt 5.15.

## darkModeGrayscale

bool, doesn't seem to have any effect (?)

default: false

## darkModeContrast

-1 to +1

default: 0

## darkModeClassifierType

- kIcon = 0
- kGeneric = 1 (default)

## darkModeImagePolicy

- kFilterAll = 0 (Apply dark-mode filter to all images.)
- kFilterNone = 1 (default: Never apply dark-mode filter to any images.)
- kFilterSmart = 2 (Apply dark-mode based on image content.)

## darkModePagePolicy

- kFilterAll = 0 (Apply dark-mode filter to all frames, regardless of content.)
- kFilterByBackground = 1 (default: Apply dark-mode filter to frames based on background color)

## darkModeTextBrightnessThreshold

```
// Text colors with brightness below this threshold will be inverted, and
// above it will be left as in the original, non-dark-mode page.  Set to 256
// to always invert text color or to 0 to never invert text color.
```

default: 150

## darkModeBackgroundBrightnessThreshold

```
// Background elements with brightness above this threshold will be inverted,
// and below it will be left as in the original, non-dark-mode page.  Set to
// 256 to never invert the color or to 0 to always invert it.
//
// Warning: This behavior is the opposite of text_brightness_threshold!
```

default: 205

## darkModeImageGrayscale

default: 0.0

---

I'll need to think how to expose all this to qutebrowser - but you can already use those settings by adding something like `blink-settings=darkMode=4,darkModeImagePolicy=0` to `qt.args`.

**Comment by naefl** [comment]
It appears that I'm not seeing anything change - I'm assuming I'm using incorrect syntax.

Here's my config.py

`c.qt.args = ["blink-settings=darkMode=4,darkModeImagePolicy=0"]`

Sourced and restarted Qutebrowser to no avail.

**Comment by PeterNerlich** [comment]
@naefl Which Qt version are you on? [I read that these are working starting Qt 5.14](https://www.reddit.com/r/qutebrowser/comments/eu93s0/dark_mode_please/ffni8dt?utm_source=share&utm_medium=web2x)

**Comment by naefl** [comment]
Yes I did read that - however I'm on 5.14.1 installed through Homebrew where the issue occurs

**Comment by The-Compiler** [comment]
@naefl If you use the .dmg, what you installed via brew is irrelevant - see `:version` inside qutebrowser.

**Comment by naefl** [comment]
`Version info
qutebrowser v1.10.0
Git commit: db05e19ae (2020-02-02 21:44:34 +0100)
Backend: QtWebEngine (Chromium 77.0.3865.129)
Qt: 5.14.1

CPython: 3.7.6
PyQt: 5.14.1

sip: 5.1.1
colorama: 0.4.3
pypeg2: 2.15
jinja2: 2.11.1
pygments: 2.5.2
yaml: 5.3
cssutils: 1.0.2 $Id$
attr: 19.3.0
PyQt5.QtWebEngineWidgets: yes
PyQt5.QtWebEngine: no
PyQt5.QtWebKitWidgets: no
pdf.js: 2.2.228 (bundled)
sqlite: 3.30.1
QtNetwork SSL: Secure Transport, macOS 10.15
`

Appears to be 5.14.1

**Comment by The-Compiler** [comment]
@naefl No idea then. Your setting above works fine for me, even on macOS. Can you run `:messages debug`, then do a search via `/Qt arguments` and show the line there?

**Comment by naefl** [comment]
@The-Compiler problem was resolved on another fresh install, flags work now. Thanks for the help!

**Comment by rien333** [comment]
@The-Compiler Yeah reverting that commit/"bug" fix seems to work, thanks. 

Just some wishful thinking, but do you think there will ever be an option to revert the old behavior? It's really surprising how much better `blink-settings=darkMode=4` works with the bug in place.

**Comment by The-Compiler** [comment]
I doubt it (maybe if/when darkmode is an official QtWebEngine setting, but I haven't opened an issue for that yet).

However, there are some more related changes to Qt 5.15, maybe they'll change that behavior again in some way:

- [Use SetBaseBackgroundColorOverride for setting background color (If4c1e228) · Gerrit Code Review](https://codereview.qt-project.org/c/qt/qtwebengine/+/296225)
- [Stop sending ViewMsg_SetBackgroundOpaque to renderer (I02bed3f0) · Gerrit Code Review](https://codereview.qt-project.org/c/qt/qtwebengine-chromium/+/296227)

**Comment by rien333** [comment]
Cool, that at least gives me hope. A future where I have to build qt5-webengine from source all the time doesn't seem so appealing, so thanks for all the info!

**Comment by D-Nice** [comment]
I share similar grievances as with @rien333. To get proper dark mode, I also had to revert webpage.bg color to white causing quite a disturbing flash on page loads.

**Comment by The-Compiler** [comment]
I now split the dark mode discussion into a separate issue (#5394) to keep things separate - let's keep this one about other interesting flags.

**Comment by The-Compiler** [comment]
FWIW I played with `--enable-features` and [content_features.cc](https://source.chromium.org/chromium/chromium/src/+/master:content/public/common/content_features.cc) a bit (trying to find a workaround for a different bug) - but I'm not sure if any of them actually work, I couldn't find anything which has an observable effect.

**Comment by The-Compiler** [comment]
I discovered and exposed `--enable-features OverlayScrollbar` (which *does* work just fine) via `scrolling.bar = overlay` (which also is the default now).

**Comment by The-Compiler** [comment]
FWIW with Qt 5.15.1 it should become possible to enable QUIC: https://codereview.qt-project.org/c/qt/qtwebengine/+/305819

**Comment by mortezadadgar** [comment]
what about `enable-accelerated-video-decode` and `enable-gpu-rasterization` which could be exposed as well?

**Comment by The-Compiler** [comment]
@mortezadadgar See #5378

**Comment by mortezadadgar** [comment]
@The-Compiler Yeah I saw this it looks like you want to enable them by default instead of exposing configs for them?
f

**Comment by The-Compiler** [comment]
@mortezadadgar Only if they actually do something useful. [According to Qt](https://bugreports.qt.io/browse/QTBUG-91677) they don't.
